spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/integration/TraceContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,203 @@
+import org.springframework.util.Assert;
+
+/**
+ * The {@link ExecutorChannelInterceptor} implementation responsible for
+ * the {@link Span} propagation from one message flow's thread to another
+ * through the {@link MessageChannel}s involved in the flow.
+ * <p>
+ * In addition this interceptor cleans up (restores) the {@link Span}
+ * in the containers Threads for channels like
+ * {@link org.springframework.integration.channel.ExecutorChannel}
+ * and {@link org.springframework.integration.channel.QueueChannel}.
+ * @author Spencer Gibb
+ * @since 1.0
+ */
+public class TraceContextPropagationChannelInterceptor
+		extends ChannelInterceptorAdapter implements ExecutorChannelInterceptor, ApplicationEventPublisherAware {
+
+	private final static ThreadLocal<Span> ORIGINAL_CONTEXT = new ThreadLocal<>();
+
+	private ApplicationEventPublisher publisher;
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher publisher) {
+		this.publisher = publisher;
+	}
+
+	@Override
+	public final Message<?> preSend(Message<?> message, MessageChannel channel) {
+		if (DirectChannel.class.isAssignableFrom(AopUtils.getTargetClass(channel))) {
+			return message;
+		}
+
+		//TODO: start span from headers?
+		Span span = getCurrentSpan();
+
+		if (span != null) {
+			publish(new ClientSentEvent(this, span));
+			return new MessageWithSpan(message, span);
+		}
+		else {
+			return message;
+		}
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public final Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		if (message instanceof MessageWithSpan) {
+			MessageWithSpan messageWithSpan = (MessageWithSpan) message;
+			Message<?> messageToHandle = messageWithSpan.message;
+			populatePropagatedContext(messageWithSpan.span, messageToHandle, channel);
+
+			publish(new ClientReceivedEvent(this, messageWithSpan.span));
+			return message;
+		}
+		return message;
+	}
+
+	@Override
+	public void afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+		Span originalContext = ORIGINAL_CONTEXT.get();
+		try {
+			if (originalContext == null) {
+				setCurrentSpan(null);
+				ORIGINAL_CONTEXT.remove();
+			}
+			else {
+				setCurrentSpan(originalContext);
+			}
+		}
+		catch (Throwable t) {//NOSONAR
+			setCurrentSpan(null);
+		}
+	}
+
+	@Override
+	public final Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		return postReceive(message, channel);
+	}
+
+	private void publish(ApplicationEvent event) {
+		if (this.publisher !=null) {
+			this.publisher.publishEvent(event);
+		}
+	}
+
+	private String getParentId(Span span) {
+		return span.getParents() != null && !span.getParents().isEmpty() ? span
+				.getParents().get(0) : null;
+	}
+
+	protected void populatePropagatedContext(Span span, Message<?> message,
+											 MessageChannel channel) {
+		if (span != null) {
+			Span currentContext = getCurrentSpan();
```

Comment:
I'm still fuzzy on what I should be doing with the span here.  This is leftover from `SecurityContextPropagationChannelInterceptor`.